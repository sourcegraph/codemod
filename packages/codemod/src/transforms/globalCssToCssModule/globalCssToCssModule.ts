import { existsSync, readFileSync, promises as fsPromises } from 'fs'
import path from 'path'

import signale from 'signale'

import { CodemodResult, CodemodOptions } from '../../types'
import { formatWithPrettierEslint, formatWithStylelint, isDefined } from '../../utils'
import { addClassNamesUtilImportIfNeeded } from '../../utils/classNamesUtility'

import { getCssModuleExportNameMap } from './postcss/getCssModuleExportNameMap'
import { transformFileToCssModule } from './postcss/transformFileToCssModule'
import { STYLES_IDENTIFIER } from './ts/processNodesWithClassName'
import { transformComponentFile } from './ts/transformComponentFile'

/**
 * Convert globally scoped stylesheet tied to the React component into a CSS Module.
 *
 * 1) Find `.tsx` file.
 * 2) Check if corresponding `.scss` file exists in the same folder.
 * 3) Convert this `.scss` file into `.module.scss`.
 * 4) Get info about CSS class names and matching export tokens.
 * 5) Replace all matching class names with export tokens.
 * 6) Add `classNames` import to the `.tsx` file if needed.
 * 7) Add `.module.scss` import to the `.tsx` file.
 *
 */
export async function globalCssToCssModule(options: CodemodOptions): CodemodResult {
    const { project, shouldWriteFiles, shouldFormat } = options
    /**
     * Find `.tsx` files with co-located `.scss` file.
     * For example `RepoHeader.tsx` should have matching `RepoHeader.scss` in the same folder.
     */
    const itemsToProcess = project
        .getSourceFiles()
        .map(tsSourceFile => {
            const tsFilePath = tsSourceFile.getFilePath()

            const parsedTsFilePath = path.parse(tsFilePath)
            const cssFilePath = path.resolve(parsedTsFilePath.dir, `${parsedTsFilePath.name}.scss`)

            if (existsSync(cssFilePath)) {
                return {
                    tsSourceFile,
                    cssFilePath,
                }
            }

            return undefined
        })
        .filter(isDefined)

    if (itemsToProcess.length === 0) {
        signale.warn('No files to process!')

        return false
    }

    const codemodResultPromises = itemsToProcess.map(async ({ tsSourceFile, cssFilePath }) => {
        const tsFilePath = tsSourceFile.getFilePath()
        const parsedTsFilePath = path.parse(tsFilePath)

        signale.info(`Processing file "${tsFilePath}"`)

        const sourceCss = readFileSync(cssFilePath, 'utf8')
        const exportNameMap = await getCssModuleExportNameMap(sourceCss)
        const { css: cssModuleSource, filePath: cssModuleFileName } = await transformFileToCssModule({
            sourceCss,
            sourceFilePath: cssFilePath,
        })

        transformComponentFile({ tsSourceFile, exportNameMap, cssModuleFileName })
        addClassNamesUtilImportIfNeeded(tsSourceFile)
        tsSourceFile.addImportDeclaration({
            defaultImport: STYLES_IDENTIFIER,
            moduleSpecifier: `./${path.parse(cssModuleFileName).base}`,
        })

        if (shouldFormat) {
            formatWithPrettierEslint(tsSourceFile)
        }
        const formattedCssModuleSource = await formatWithStylelint(cssModuleSource, cssFilePath)

        /**
         * If `shouldWriteFiles` is true:
         *
         * 1. Update TS file with a new source that uses CSS module.
         * 2. Create a new CSS module file.
         * 3. Delete redundant SCSS file that's replaced with CSS module.
         */
        const fsWritePromise = shouldWriteFiles
            ? Promise.all([
                  tsSourceFile.save(),
                  fsPromises.writeFile(cssModuleFileName, formattedCssModuleSource, { encoding: 'utf-8' }),
                  fsPromises.rm(cssFilePath),
              ])
            : undefined

        return {
            fsWritePromise,
            css: {
                source: formattedCssModuleSource,
                path: path.resolve(parsedTsFilePath.dir, cssModuleFileName),
            },
            ts: {
                source: tsSourceFile.getFullText(),
                path: tsSourceFile.getFilePath(),
            },
        }
    })

    return Promise.all(codemodResultPromises)
}
