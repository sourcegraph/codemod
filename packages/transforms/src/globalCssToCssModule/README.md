# Global CSS to CSS Module codemod

Convert globally scoped stylesheet tied to the React component into a CSS Module.

1. Find `.tsx` file.
2. Check if corresponding `.scss` file exists in the same folder.
3. Convert this `.scss` file into `.module.scss`.
4. Get info about CSS class names and matching CSS module export tokens.
5. Replace all matching class names with export tokens.
6. Add `classNames` import to the `.tsx` file if needed.
7. Add `.module.scss` import to the `.tsx` file.

## Usage

See available commands and options:

```sh
yarn transform --help
```

Apply codemod:

```sh
yarn transform -t globalCssToCssModule --write true "/client/web/src/search/**/*.tsx"
```

To automatically fix formatting in the updated React components use the `--format true` flag.

```sh
yarn transform -t globalCssToCssModule --write true --format true "/client/web/src/search/**/*.tsx"
```

Currently, it's quite slow because a new ESLint instance is initialized for each file. Consider running `eslint --fix` after the codemod if you're changing a lot of files in one command.

## Post codemod steps for the Sourcegraph repo

1. Remove global styles import from the codebase: `@import 'DeletedGlobalStylesFile';`.
2. Check "Unused CSS classes" reported in the CLI output. These are classes that exist in the CSS module but were not found in the adjacent React component.
   1. Remove redundant CSS classes from the CSS module.
   2. Fix classes used outside of the component by using CSS module there or convert them to props to eliminate the implicit dependency.
3. Double-check the usage of the deleted global CSS classes in integration tests or other stylesheets.
   1. If global CSS class is used in the integration test — add `data-testid="some-name-here"` to the relevant Jsx element and use `[data-testid="some-name-here"]` selector in the integration test.
4. Fix usage of the global SCSS variables in the CSS module:`color: $success; -> color: var(--success);`
   1. If CSS variable doesn't exist — add it next to the SCSS variable: `$spacer: 1px; -> $spacer: 1px; :root { --spacer: #{$spacer}; }`
5. Double-check that the `:global(...)` selector usage in the created CSS module is correct.
6. Run `yarn test -u` to update Jest snapshots.
7. Run `eslint --fix` to update the Typescript source files formatting if `--format` flag was not used in the codemod command.
