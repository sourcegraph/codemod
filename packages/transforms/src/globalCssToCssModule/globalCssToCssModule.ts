import path from 'path'

import signale from 'signale'

import { Codemod } from '@sourcegraph/codemod-cli'
import { isDefined } from '@sourcegraph/codemod-common'
import { formatWithStylelint } from '@sourcegraph/codemod-toolkit-css'
import { utilities } from '@sourcegraph/codemod-toolkit-packages'
import { formatWithPrettierEslint } from '@sourcegraph/codemod-toolkit-ts'

import { getCssModuleExportNameMap } from './postcss/getCssModuleExportNameMap'
import { transformFileToCssModule } from './postcss/transformFileToCssModule'
import { STYLES_IDENTIFIER } from './ts/processNodesWithClassName'
import { transformComponentFile } from './ts/transformComponentFile'

/**
 * Convert globally scoped stylesheet tied to the React component into a CSS Module.
 *
 * 1) Find `.tsx` file.
 * 2) Check if corresponding `.scss` file exists in the same folder.
 * 3) Convert this `.scss` file into `.module.scss`.
 * 4) Get info about CSS class names and matching export tokens.
 * 5) Replace all matching class names with export tokens.
 * 6) Add `classNames` import to the `.tsx` file if needed.
 * 7) Add `.module.scss` import to the `.tsx` file.
 *
 */
export const globalCssToCssModule: Codemod = context => {
    const { project, shouldWriteFiles, shouldFormat, classname } = context
    const fs = project.getFileSystem()

    /**
     * Find `.tsx` files with co-located `.scss` file.
     * For example `RepoHeader.tsx` should have matching `RepoHeader.scss` in the same folder.
     */
    const itemsToProcess = project
        .getSourceFiles()
        .map(tsSourceFile => {
            const tsFilePath = tsSourceFile.getFilePath()

            const parsedTsFilePath = path.parse(tsFilePath)

            if (parsedTsFilePath.ext !== '.tsx') {
                return
            }

            const cssFilePath = path.resolve(parsedTsFilePath.dir, `${parsedTsFilePath.name}.scss`)

            if (fs.fileExistsSync(cssFilePath)) {
                return {
                    tsSourceFile,
                    cssFilePath,
                }
            }

            return undefined
        })
        .filter(isDefined)

    if (itemsToProcess.length === 0) {
        signale.warn('No files to process!')

        return Promise.resolve([])
    }

    const codemodResultPromises = itemsToProcess.map(async ({ tsSourceFile, cssFilePath }) => {
        const tsFilePath = tsSourceFile.getFilePath()
        const parsedTsFilePath = path.parse(tsFilePath)

        signale.info(`Processing file "${tsFilePath}"`)

        const sourceCss = fs.readFileSync(cssFilePath, 'utf8')
        const exportNameMap = await getCssModuleExportNameMap(sourceCss)
        const { css: cssModuleSource, filePath: cssModuleFileName } = await transformFileToCssModule({
            sourceCss,
            sourceFilePath: cssFilePath,
        })

        transformComponentFile({ tsSourceFile, exportNameMap, cssModuleFileName, classname })

        utilities[classname].importer(tsSourceFile)

        tsSourceFile.addImportDeclaration({
            defaultImport: STYLES_IDENTIFIER,
            moduleSpecifier: `./${path.parse(cssModuleFileName).base}`,
        })

        if (shouldFormat) {
            formatWithPrettierEslint(tsSourceFile)
        }
        const formattedCssModuleSource = await formatWithStylelint(cssModuleSource, cssFilePath)

        /**
         * If `shouldWriteFiles` is true:
         *
         * 1. Update TS file with a new source that uses CSS module.
         * 2. Create a new CSS module file.
         * 3. Delete redundant SCSS file that's replaced with CSS module.
         */
        const fsWritePromise = shouldWriteFiles
            ? Promise.all([
                  tsSourceFile.save(),
                  fs.writeFile(cssModuleFileName, formattedCssModuleSource),
                  fs.delete(cssFilePath),
              ])
            : undefined

        return {
            target: tsSourceFile,
            manualChangesReported: {},
            fsWritePromise,
            files: [
                {
                    source: formattedCssModuleSource,
                    path: path.resolve(parsedTsFilePath.dir, cssModuleFileName),
                },
                {
                    source: tsSourceFile.getFullText(),
                    path: tsSourceFile.getFilePath(),
                },
            ],
        }
    })

    return Promise.all(codemodResultPromises)
}
