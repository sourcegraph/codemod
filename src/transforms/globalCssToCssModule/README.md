# Global CSS to CSS Module codemod

Convert globally scoped stylesheet tied to the React component into a CSS Module.

1. Find `.tsx` file.
2. Check if corresponding `.scss` file exists in the same folder.
3. Convert this `.scss` file into `.module.scss`.
4. Get info about CSS class names and matching CSS module export tokens.
5. Replace all matching class names with export tokens.
6. Add `classNames` import to the `.tsx` file if needed.
7. Add `.module.scss` import to the `.tsx` file.

## Usage

```sh
yarn transform -t globalCssToCssModule --write true "/client/web/src/search/**/*.tsx"
```
