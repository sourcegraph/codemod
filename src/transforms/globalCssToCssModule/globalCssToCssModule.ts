import { existsSync, readFileSync, promises as fsPromises } from 'fs'
import path from 'path'

import { Project } from 'ts-morph'

import { isDefined } from '../../utils'

import { getCssModuleExportNameMap } from './postcss/getCssModuleExportNameMap'
import { transformFileToCssModule } from './postcss/transformFileToCssModule'
import { addClassNamesUtilImportIfNeeded } from './ts/classNamesUtility'
import { getNodesWithClassName } from './ts/getNodesWithClassName'
import { STYLES_IDENTIFIER, processNodesWithClassName } from './ts/processNodesWithClassName'

interface GlobalCssToCssModuleOptions {
    project: Project
    /** If `true` persist changes made by the codemod to the filesystem. */
    shouldWriteFiles?: boolean
}

interface CodemodResult {
    css: {
        source: string
        path: string
    }
    ts: {
        source: string
        path: string
    }
    fsWritePromise?: Promise<void[]>
}

/**
 * Convert globally scoped stylesheet tied to the React component into a CSS Module.
 *
 * 1) Find `.tsx` file.
 * 2) Check if corresponding `.scss` file exists in the same folder.
 * 3) Convert this `.scss` file into `.module.scss`.
 * 4) Get info about CSS class names and matching export tokens.
 * 5) Replace all matching class names with export tokens.
 * 6) Add `classNames` import to the `.tsx` file if needed.
 * 7) Add `.module.scss` import to the `.tsx` file.
 *
 */
export async function globalCssToCssModule(options: GlobalCssToCssModuleOptions): Promise<CodemodResult[]> {
    const { project, shouldWriteFiles } = options
    /**
     * Find `.tsx` files with co-located `.scss` file.
     * For example `RepoHeader.tsx` should have matching `RepoHeader.scss` in the same folder.
     */
    const itemsToProcess = project
        .getSourceFiles()
        .map(tsSourceFile => {
            const tsFilePath = tsSourceFile.getFilePath()

            const parsedTsFilePath = path.parse(tsFilePath)
            const cssFilePath = path.resolve(parsedTsFilePath.dir, `${parsedTsFilePath.name}.scss`)

            if (existsSync(cssFilePath)) {
                return {
                    tsSourceFile,
                    cssFilePath,
                }
            }

            return undefined
        })
        .filter(isDefined)

    const codemodResultPromises = itemsToProcess.map(async ({ tsSourceFile, cssFilePath }) => {
        const tsFilePath = tsSourceFile.getFilePath()
        const parsedTsFilePath = path.parse(tsFilePath)

        const sourceCss = readFileSync(cssFilePath, 'utf8')
        const { css: cssModuleSource, filePath: cssModuleFileName } = await transformFileToCssModule({
            sourceCss,
            sourceFilePath: cssFilePath,
        })
        const exportNameMap = await getCssModuleExportNameMap(cssModuleSource)

        processNodesWithClassName({
            exportNameMap,
            nodesWithClassName: getNodesWithClassName(tsSourceFile),
        })

        addClassNamesUtilImportIfNeeded(tsSourceFile)
        tsSourceFile.addImportDeclaration({
            defaultImport: STYLES_IDENTIFIER,
            moduleSpecifier: `./${path.parse(cssModuleFileName).base}`,
        })

        /**
         * If `shouldWriteFiles` is true:
         *
         * 1. Update TS file with a new source that uses CSS module.
         * 2. Create a new CSS module file.
         * 3. Delete redundant SCSS file that's replaced with CSS module.
         */
        const fsWritePromise = shouldWriteFiles
            ? Promise.all([
                  tsSourceFile.save(),
                  fsPromises.writeFile(cssModuleFileName, cssModuleSource, { encoding: 'utf-8' }),
                  fsPromises.rm(cssFilePath),
              ])
            : undefined

        return {
            fsWritePromise,
            css: {
                source: cssModuleSource,
                path: path.resolve(parsedTsFilePath.dir, cssModuleFileName),
            },
            ts: {
                source: tsSourceFile.getFullText(),
                path: tsSourceFile.getFilePath(),
            },
        }
    })

    const codemodResults = await Promise.all(codemodResultPromises)

    if (shouldWriteFiles) {
        await Promise.all(codemodResults.map(result => result.fsWritePromise))
    }

    return codemodResults
}
