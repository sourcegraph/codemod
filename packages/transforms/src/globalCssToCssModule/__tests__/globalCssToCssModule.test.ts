import path from 'path'

import signale from 'signale'
import { Project } from 'ts-morph'

import { globalCssToCssModule } from '../globalCssToCssModule'

const TARGET_FILE = path.resolve(__dirname, './fixtures/MyComponent.tsx')

describe('globalCssToCssModule', () => {
    beforeEach(() => {
        signale.disable()
    })

    it('transforms correctly', async () => {
        const project = new Project()
        project.addSourceFilesAtPaths(TARGET_FILE)
        const [{ files }] = await globalCssToCssModule({ project, shouldFormat: true })

        expect(files).toBeTruthy()

        if (files) {
            const [cssModule, reactComponent] = files

            expect(cssModule.source).toMatchSnapshot()
            expect(reactComponent.source).toMatchSnapshot()
        }
    }, 15000) // Timeout of 15s (default is 5s). `prettier-eslint` format is slow ðŸ˜¬.
})
