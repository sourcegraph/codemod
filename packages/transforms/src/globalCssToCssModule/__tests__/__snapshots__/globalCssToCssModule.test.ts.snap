// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`globalCssToCssModule transforms correctly 1`] = `
"

/* Woah!*/
.kek {
    color: red;

    /* nah*/
    &--wow {
        color: blue;
    }
}

/* Let's see how you handle global top-level classes!*/
:global(.theme-light) .kek {
    color: red;
}

/* .repo-header {*/
.repo-header {
    flex: none;
    padding: 0;
    align-items: stretch;
    border-color: var(--border-color);
    border-style: solid;

    /* &--alert {*/
    &--alert {
        border-width: 1px 0;
    }

    /* .navbar-nav {*/
    :global(.navbar-nav) {
        white-space: nowrap;

        :global(.nav-item) {
            display: flex;
            align-items: stretch;

            /* .nav-link {*/
            :global(.nav-link) {
                user-select: none;
            }
        }
    }

    :global(.copy-link-action) {
        opacity: 0;
    }

    &:hover,
    &:focus-within {
        :global(.copy-link-action) {
            opacity: 1;
        }
    }
}

.logo {
    display: flex;
}

/* &__action-list-item {*/
.action-list-item {
    /* Have a small gap between buttons so they are visually distinct when pressed*/
    /* stylelint-disable-next-line declaration-property-unit-whitelist*/
    margin-left: 1px;

    /* &:hover {*/
    &:hover {
        /* background: var(--color-bg-1);*/
        background: var(--color-bg-1);

        /* .theme-light & {*/
        :global(.theme-light) & {
            background: inherit;
        }
    }
}

/* &__file-action {*/
.file-action {
    display: block;
    text-align: left;

    /* span {*/
    span {
        color: var(--body-color);
        margin-left: 0.5rem;
    }
}

/* &__action {*/
.action {
    margin: 0.5rem 0.625rem 0.5rem 0;
    padding: 0.25rem;

    :global(.theme-light) & {
        margin-top: 0;
        margin-bottom: 0;
    }
}

.alert {
    margin: 0 0.25rem;
    padding: 0.125rem 0.25rem;
    cursor: default;
    user-select: none;
}

/* &__icon-chevron {*/
.icon-chevron {
    opacity: 0.6;
    margin: auto;
}

.spacer {
    flex: 1 1 0;
}

/* &__kek-pek {*/
.kek-pek {
    color: red;
}
"
`;

exports[`globalCssToCssModule transforms correctly 2`] = `
"// @ts-nocheck
import React from 'react'
import classNames from 'classnames'
import styles from './MyComponent.module.scss'

export const MyComponent = () => {
    const isActive = true

    return (
        <div className={classNames('d-flex m-1', styles.kek, styles.kekWow)}>
            <div
                className={classNames('m-1', true ? styles.kek : false, 'd-flex', {
                    [styles.kek]: false,
                    [classNames('d-flex mr-1', styles.kek)]: isActive,
                })}
            ></div>
            It's a component<p className={styles.logo}>wow</p>
            <div className={classNames('m-2 d-flex m-1', styles.repoHeader)}>Another one!</div>
        </div>
    )
}
"
`;
